src/app/EventManagement.cpp

```diff
@@ -748,6 +751,71 @@ CHIP_ERROR EventManagement::FetchEventsSince(TLVWriter & aWriter, const ObjectLi
     return err;
 }
 
+CHIP_ERROR EventManagement::RemoveInvalidFabricCB(const TLV::TLVReader & aReader, size_t aDepth, void * apContext)
+{
+    // the function does not actually remove the event, instead, it sets the fabric index to an invalid value.
+    FabricIndex * invalidFabricIndex = static_cast<FabricIndex *>(apContext);
+
+    TLVReader event;
+    TLVType tlvType;
+    TLVType tlvType1;
+    event.Init(aReader);
+    ReturnErrorOnFailure(event.EnterContainer(tlvType));
+    ReturnErrorOnFailure(event.Next());
+    ReturnErrorOnFailure(event.EnterContainer(tlvType1));
+    ReturnErrorOnFailure(event.Next());
+
+    uint8_t fabricIndex = 0;
+    while (CHIP_NO_ERROR == event.Next())
+    {
+        if (event.GetTag() == TLV::ProfileTag(kEventManagementProfile, kFabricIndexTag))
+        {
+            ReturnErrorOnFailure(event.Get(fabricIndex));
+            if (fabricIndex == *invalidFabricIndex)
+            {
+                // fabricIndex is encoded as an integer; the dataPtr will point to a location immediately after its encoding
+                uint8_t * dataPtr                  = const_cast<uint8_t *>(event.GetReadPoint());
+                CHIPCircularTLVBuffer * readBuffer = static_cast<CHIPCircularTLVBuffer *>(event.GetBackingStore());
+                // shift the dataPtr to point to the encoding of the fabric index, accounting for wraparound in backing storage
+                if (readBuffer->GetQueue() != dataPtr)
+                {
+                    dataPtr = dataPtr - 1;
+                }
+                else
+                {
+                    dataPtr = readBuffer->GetQueue() + readBuffer->GetTotalDataLength() - 1;
+                }
+                // set the value of fabricIndex to 0
+                memset(dataPtr, 0, 1);
+                event.Get(fabricIndex);
+                return CHIP_NO_ERROR;
+            }
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EventManagement::RemoveInvalidFabric(FabricIndex aFabricIndex)
+{
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+    const bool recurse = false;
+    TLVReader reader;
+    CircularEventBufferWrapper bufWrapper;
+
+#if !CHIP_SYSTEM_CONFIG_NO_LOCKING
+    ScopedLock lock(sInstance);
+#endif // !CHIP_SYSTEM_CONFIG_NO_LOCKING
```

Comment:
Wait, why is EventManagement doing internal synchronization?  Followup to stop doing that, please....

Reply:
in the past, I was worried that it would be used by multiple thread, now I think we can remove this.

Reply:
https://github.com/project-chip/connectedhomeip/issues/18065